### PR TITLE
Allocate HIR on an arena 3/4 -- Ty

### PR DESCRIPTION
**Pull request by cjgillot**
This is the third PR in the series started by #66931 and #66936

Once again, commits don't really make sense on their own.
They are mostly split by type of compile error.

The additional diff is here: https://github.com/cjgillot/rust/compare/hirene-expr...hirene-ty

### CODE REVIEW

**Comment by davidtwco** [comment]
r? @eddyb
cc @Zoxc

File: src/librustc/hir/lowering/item.rs [hunk 1]
```diff
@@ -1444,13 +1436,13 @@ impl<'hir> LoweringContext<'_, 'hir> {
                     &bound_generic_params,
                     |this| {
                         hir::WherePredicate::BoundPredicate(hir::WhereBoundPredicate {
-                            bound_generic_params: this.lower_generic_params(
+                            bound_generic_params: this.arena.alloc_from_iter(this.lower_generic_params(
                                 bound_generic_params,
                                 &NodeMap::default(),
                                 ImplTraitContext::disallowed(),
-                            ),
+                            ).into_iter()),
```
Comment (Zoxc) [inline src/librustc/hir/lowering/item.rs]: `lower_generic_params` could probably return an arena slice.

File: src/librustc/hir/lowering.rs [hunk 2]
```diff
@@ -946,7 +946,7 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
         let mut lowered_params: Vec<_> = lowered_generics
             .params
             .into_iter()
-            .chain(in_band_defs)
+            .chain(in_band_defs.into_iter())
```
Comment (Zoxc) [inline src/librustc/hir/lowering.rs]: Why is this needed?

File: src/librustc/hir/lowering.rs [hunk 3]
```diff
@@ -2057,20 +2063,24 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
             segment.ident, segment.id, id,
         );
 
-        hir::PathSegment::new(
```
Comment (Zoxc) [inline src/librustc/hir/lowering.rs]: Was this the only use of `new`?

File: src/librustc/hir/lowering.rs [hunk 4]
```diff
@@ -2740,21 +2748,27 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
 
-    fn lower_param_bounds(&mut self, bounds: &[GenericBound], mut itctx: ImplTraitContext<'_>)
-                          -> hir::GenericBounds<'hir> {
+    fn lower_param_bounds(&mut self, bounds: &[GenericBound], itctx: ImplTraitContext<'_, 'hir>)
+        -> hir::GenericBounds<'hir>
+    {
+        self.arena.alloc_from_iter(self.lower_param_bounds_mut(bounds, itctx))
+    }
+
+    fn lower_param_bounds_mut(&mut self, bounds: &[GenericBound], mut itctx: ImplTraitContext<'_, 'hir>)
```
Comment (Zoxc) [inline src/librustc/hir/lowering.rs]: This should probably return an iterator instead of a `Vec`.

File: src/librustc/hir/lowering.rs (hunk 3 above)
Comment (cjgillot) [reply src/librustc/hir/lowering.rs]: Yes. It has been removed.

**Comment by cjgillot** [comment]
Rebased with review comments.

**Comment by matthewjasper** [comment]
Bors restart added this to the queue again.
@bors r-
